flight/Libraries/math/smoothcontrol.c

```diff
@@ -0,0 +1,195 @@
+#include <stdio.h>
+#include <stdint.h>
+#include <math.h>
+
+#include "openpilot.h"
+#include "pios.h"
+#include "smoothcontrol.h"
+
+#define ASSERT(x) if(x) PIOS_Assert(0);
+#define MAX(v,x) ((v) > x ? x : (v))
+
+struct smoothcontrol_axis_state {
+	// Last known signal value. Needed to detect change.
+	float signal;
+
+	// Current state of axis.
+	float current;
+
+	// Amount of differential per step of integration.
+	float differential;
+
+	// Stores a time-out when to stop predicting. Unit is integrator ticks.
+	uint8_t integrator_timeout;
+
+	// Which conditioning algorithm.
+	uint8_t mode;
+
+	// Track integration time.
+	uint8_t tick_counter;
+
+	// Stores the lowest amount of ticks between control input updates.
+	// uint8_t allows for 32ms at 8KHz, should be sufficient to deal with
+	// even PPM and PWM at that rates (plz no).
+	uint8_t control_interval;
+};
+
+struct smoothcontrol_state_internal {
+
+	// Maximum ticks to integrate. Calculate value from dT and constant. Unit is ticks.
+	uint8_t time_bomb;
+
+	// RPY+Thrust.
+	struct smoothcontrol_axis_state axis[4];
+};
+
+// Ran on signal change, recalculates the diffs and timeouts for the specific mode.
+static void smoothcontrol_update(smoothcontrol_state state, struct smoothcontrol_axis_state *axis, const float new_signal)
+{
+	float signal_diff = new_signal - axis->signal;
+
+	switch(axis->mode) {
+		default:
+			// Override any crap values.
+			axis->mode = SMOOTHCONTROL_NONE;
+		case SMOOTHCONTROL_NONE:
+			// Doesn't do anything in the integrator.
+			break;
+
+		// icee's proposal, creates chamfered steps with signal prediction
```

Comment:
We need a worst-case timeout that applies no matter what.  On the order of 15ms.